### ISSUE TITLE
torch.Tensor() is stuck or deadlock in multiprocess in torch 1.2.0

### ISSUE THREAD

**Issue by schen119** [pytorch/pytorch#26447]
## 🐛 Bug
The ```torch.Tensor()``` will be stuck or deadlock in the multiprocess if the main process did the  ```torch.Tensor()``` before the child-process start().
This bug is not happened in torch 1.1.0. Please see the code below to reproduce the bug.
 If reducing the size of dummy_im to the size (3,3), it's ok. Therefore, It might be a shared memory problem.

## To Reproduce
```
import numpy as np
import torch
import sys
import torch.multiprocessing as mp


if __name__ == '__main__':
    print("Python version=", sys.version)  # Python3.4.4
    print("torch version=", torch.__version__) # torch 1.2.0
    image_queue = mp.Queue(50)

    #########################################
    #comment two lines below and no deadlock, otherwise deadlock in line 23
    dummy_im = np.random.randn(300, 400)
    dummy_im = torch.Tensor(dummy_im)
    #########################################

    def _read_worker():
        im = image_queue.get(block=True)
        print(type(im))
        print(im.shape)
        # stuck here: deadlock? line 23
        im = torch.Tensor(im)
        print(im)

    w = mp.Process(target=_read_worker)
    w.daemon = True  # ensure that the worker exits on process exit
    w.start()
    im = np.random.randn(300, 400)
    image_queue.put(im, block=True)
    w.join()
    print("done")
```

- stuck in the line ```im = torch.Tensor(im)``` with no exception error message.

- Reduce the dummy_im to the size (3,3), it becomes work.

- if move the ```dummy_im = torch.Tensor(dummy_im)``` after ```w.start()```, it's ok.

- My torch version is 1.2.0 but if I downgrade to 1.1.0, everything is ok.

## Expected behavior
The normal expected result in torch 1.1.0:
```
Python version= 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0]
torch version= 1.1.0
<class 'numpy.ndarray'>
(300, 400)
tensor([[-0.8363,  1.3835, -0.7603,  ..., -0.0222, -0.4916, -1.2512],
        [-0.2285, -0.6330,  0.4084,  ...,  0.9458,  1.0684,  0.3801],
        [-0.5260,  1.4498, -0.8997,  ...,  0.2333, -0.8163,  1.1754],
        ...,
        [-0.2710,  0.7194, -0.2842,  ...,  0.5777,  1.1365, -0.1534],
        [-0.6067,  0.1561, -1.2539,  ...,  1.6007,  0.4443, -2.6501],
        [ 0.1211, -1.1745, -0.0238,  ...,  1.0563,  1.2978, -2.9615]])
done
```
The stuck output in torch 1.2.0:
```
Python version= 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]
torch version= 1.2.0
<class 'numpy.ndarray'>
(300, 400)
```

## Environment
 - PyTorch Version (e.g., 1.0): 1.2.0
 - OS (e.g., Linux): Ubuntu 16.04 LTS
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version:3.7.3
 - CUDA/cuDNN version:10.0
 - GPU models and configuration:Geforce GTX 1080Ti
 - Any other relevant information:

**Comment by mrshenli** [comment]
I cannot reproduce the deadlock. Have you tried the following:

1. Use `multiprocessing.queues.SimpleQueue` as suggested in [this tutorial](https://pytorch.org/docs/stable/notes/multiprocessing.html).
2. Create the tensor on the main process, move it to shared memory, and pass it to the child process through the `torch.multiprocess.Queue`.

**Comment by schen119** [comment]
@mrshenli 
1. Using ```multiprocessing.queues.SimpleQueue``` has some problem.
```
import numpy as np
import torch
import sys
import torch.multiprocessing as mp


if __name__ == '__main__':
    print("Python version=", sys.version)  # Python3.4.4
    print("torch version=", torch.__version__) # torch 1.2.0


    #image_queue = mp.Queue(50)
    image_queue = mp.SimpleQueue()
    #########################################
    # #comment two lines below and no deadlock, otherwise deadlock in line 23
    dummy_im = np.random.randn(300, 400)
    dummy_im = torch.Tensor(dummy_im)
    #########################################

    def _read_worker():
        #im = image_queue.get(block=True)
        im = image_queue.get()
        print(type(im))
        print(im.shape)
        # stuck here: deadlock?
        im = torch.Tensor(im)
        print(im)

    w = mp.Process(target=_read_worker)
    w.daemon = True  # ensure that the worker exits on process exit
    w.start()
    im = np.random.randn(300, 400)
    #image_queue.put(im, block=True)
    image_queue.put(im)
    w.join()
    print("done")
```

2. My original scenario is to initialize the model in the main process where has some ToTensor() steps and fork the process to handle the image preprocessing. The last step of the image preprocessing is ToTensor(). Then I found this issue. The scenario is different so I don't need to  ```Create the tensor on the main process, move it to shared memory, and pass it to the child process through the torch.multiprocess.Queue``` as you said.

3. What is your environment? Did you try it on any of the environments below:
PyTorch 1.1.0+ Python version:3.7.4 created by conda
PyTorch 1.2.0+ Python version:3.7.4 created by conda
PyTorch 1.2.0+ Python version:3.6.8 created by conda
PyTorch 1.2.0+ Python version:3.5.2 in Docker

**Comment by mrshenli** [comment]
> What is your environment? 

I compiled from source, the following is my environment:

```
PyTorch version: 1.3.0a0+289830a
Is debug build: No
CUDA used to build PyTorch: 9.2.88

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-36)
CMake version: version 3.14.0

Python version: 3.7
```

**Comment by andreaskoepf** [comment]
I also could not reproduce the deadlock with `Python version= 3.7.4 (default, Aug 13 2019, 20:35:49)`. I tried with Pytorch 1.2 & 1.3.0a0+54bfaf8. Both times the script ran to "done" and exited without error.

My environment:

```
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
```

**Comment by schen119** [comment]
We ran this code using a conda environment. Is it possible to limits the shared memory size which might cause this deadlock?

**Comment by mruberry** [comment]
I can reproduce this on my devfair using the original script. The workarounds described also get around the hang for me.

**Comment by mruberry** [comment]
This is not a Numpy issue. Renaming, removing Module : Numpy, and removed the "triaged" label because the actual issue is different than what was originally (reasonably) suspected.

The following code, simplified from the above, will hang:

```
import torch
import torch.multiprocessing as mp

def map_fn():
    t1 = torch.randn(300, 300) + torch.randn(300, 300)

if __name__ == '__main__':
    t0 = torch.randn(300, 300) + torch.randn(300, 300)

    w = mp.Process(target=map_fn)
    w.daemon = True  # ensure that the worker exits on process exit
    w.start()
    w.join()
```

The hang appears to be in TensorIterator. The original code was triggering this behavior because it was invoking copy kernels (that use TensorIterator) behind the scenes.

If the code is rewritten to use `mp.spawn()`, as PyTorch's documentation recommends, then no hang occurs.

**Comment by albanD** [comment]
Added triage-review to discuss all the fork vs spawn related issues.
We have a nice detection mechanism for cuda, but the autograd can deadlock as well (see https://github.com/pytorch/pytorch/issues/32835). Maybe we should have a better overall solution for this.

**Comment by albanD** [comment]
@ezyang says that the parallel_for should raise a nice error message when it happens

**Comment by ezyang** [comment]
It should be implemented the same way we detect bad cuda forks

**Comment by albanD** [comment]
Given that the same things needs to be done for the autograd (see this issue https://github.com/pytorch/pytorch/issues/32835), should we have a centralized way to detect that we're in a fork that we can check from different places (cuda init, parallel_for and autograd) ?

**Comment by ezyang** [comment]
Yes
